src/core/src/pattern/op/pattern.cpp

```diff
@@ -91,136 +84,314 @@ PatternValueMap as_pattern_value_map(const PatternMap& pattern_map) {
+        },
+        "type_matches(" + type.to_string() + ")");
 }
 
-std::function<bool(Output<Node>)> all_of(const std::vector<std::function<bool(Output<Node>)>>& predicates) {
-    return [=](Output<Node> output) -> bool {
-        for (auto& p : predicates) {
-            if (!p(output))
-                return false;
-        }
-        return true;
-    };
-}
-}  // namespace pattern
-}  // namespace pass
-pass::pattern::op::ValuePredicate operator||(const pass::pattern::op::ValuePredicate& a,
-                                             const pass::pattern::op::ValuePredicate& b) {
-    return [=](const Output<Node>& value) {
-        return a(value) || b(value);
-    };
+op::Predicate type_matches_any(const std::vector<element::Type>& expected_types) {
+    std::string arg;
+    for (size_t i = 0; i < expected_types.size(); ++i)
+        arg += (i ? "," : "") + expected_types[i].to_string();
+    return op::Predicate(
+        [=](const Output<Node>& output) -> bool {
+            const auto& output_type = output.get_element_type();
+            return std::any_of(expected_types.begin(), expected_types.end(), [=](element::Type type) {
+                return type == output_type;
+            });
+        },
+        "type_matches_any({" + arg + "})");
 }
 
-pass::pattern::op::ValuePredicate operator||(const std::function<bool(Output<Node>)>& a,
-                                             const std::function<bool(Output<Node>)>& b) {
-    return [=](const Output<Node>& value) {
-        return a(value) || b(value);
-    };
+op::Predicate all_of(const std::vector<std::function<bool(Output<Node>)>>& predicates) {
+    return op::Predicate(
+        [=](const Output<Node>& output) -> bool {
+            for (auto& p : predicates) {
+                if (!p(output))
+                    return false;
+            }
+            return true;
+        },
+        "all_of(...)");
 }
 
-pass::pattern::op::ValuePredicate operator||(const pass::pattern::op::NodePredicate& a,
-                                             const pass::pattern::op::NodePredicate& b) {
-    return pass::pattern::op::as_value_predicate(a) || pass::pattern::op::as_value_predicate(b);
-}
-
-pass::pattern::op::ValuePredicate operator||(const pass::pattern::op::ValuePredicate& a,
-                                             const pass::pattern::op::NodePredicate& b) {
-    return a || pass::pattern::op::as_value_predicate(b);
+namespace {
+bool ends_with(std::string_view str, std::string_view suffix) {
+    return str.size() >= suffix.size() && str.compare(str.size() - suffix.size(), suffix.size(), suffix) == 0;
 }
 
-pass::pattern::op::ValuePredicate operator||(const pass::pattern::op::NodePredicate& a,
-                                             const pass::pattern::op::ValuePredicate& b) {
-    return pass::pattern::op::as_value_predicate(a) || b;
+std::pair<std::vector<std::pair<int64_t, std::string>>, int64_t> parse_string(const std::string& s_global) {
```

Comment:
`parse_string` is quite common and uninformative function name
could change it? and add a short description about usage